### DISCUSSION TITLE
Use `@formatjs/intl`?

### DISCUSSION THREAD

**Discussion by longlho** [ember-intl/ember-intl#1504] [Ideas]
👋 We just released @formatjs/intl that detaches the imperative API from the React binding (react-intl uses this basically) so I'm curious if ember-intl can use it? It basically has everything that react-intl supports (custom formats, default timeZone, format merging, handling AST vs regular messages).

I'm curious what the roadmap for ember-intl is as well?

**Comment by longlho** [comment]
@jasonmit it looks like `fallbackLocale` behave differently compared to `defaultLocale` in formatjs. `defaultLocale` is basically meant to dictate the locale of the `defaultMessage`, so that when we encounter an untranslated message, everything is formatted in the same locale. This avoids scenarios where u default to English message but have Japanese number formatting basically. Looking through the code I don't see `fallbackLocale` being used at runtime? Am I misunderstanding this?

**Comment by jasonmit** [comment]
`fallbackLocale` is used only at build time.  If we discover a locale doesn't implement a translation key, it will take it from the fallback locale.

**Comment by longlho** [comment]
@jasonmit how come helpers can set its own `locale`? How do u maintain the same locale across the app?

**Comment by jasonmit** [comment]
@longlho that's optional, if one isn't provided it uses the locale from the application-wide Intl service:
https://github.com/ember-intl/ember-intl/blob/master/addon/services/intl.js#L33

**Comment by longlho** [comment]
@jasonmit What's the current state of toolchain for `ember-intl`? Are there message extraction CLI? Message AST compilation during build? Linter?

Also I see references to `intl.set('formats')` but I cannot find any doc on it?

**Comment by jasonmit** [comment]
> Are there message extraction CLI?

No

> Message AST compilation during build?

No but we can support this with a little effort

> Linter?

Yes but it could definitely be improved https://github.com/ember-intl/ember-intl/tree/master/lib/broccoli/translation-reducer/linter

**Reply by longlho** [reply]
hmm does ember ecosystem not use `eslint`?

**Comment by jasonmit** [comment]
Yes the project and community use eslint

**Reply by longlho** [reply]
Oh so the broccoli linter is more like a build plugin than a IDE linter?

**Reply by jasonmit** [reply]
@longlho correct, it runs at build time because we're aggregating ember addons at build time, which may contain ember-intl translations, and then we process those translations together with the host app to detect conflicts between translation arguments across locales and missing keys across locales.

**Comment by longlho** [comment]
@jasonmit I'm not sure I understand the usage of `t()` with `default` stuff. Seems like it's undocumented? I guess my real question is if it is used?

**Reply by jasonmit** [reply]
Is this what you're referring to? https://github.com/ember-intl/ember-intl/blob/e4a4563b40cea8a78813b720f886975ea342fa44/tests/dummy/app/pods/docs/guide/translating-text/template.md#fallback-translation

**Reply by longlho** [reply]
@jasonmit  ahh yup my searchfoo wasn't very good. Is it used? I haven't seen this use case in the React/Vue/Angular ecosystem basically.

Some translation vendors give u back just English for untranslated strings basically, which means fallback chain never gets triggered.

**Reply by jasonmit** [reply]
I imagine some do use it but don't have the data/analytics to determine how much it is used.  As long as there is a migration path to a sane alternative to accomplish similar behavior, I'd be fine with removing it.

**Reply by longlho** [reply]
I guess this is a build issue more than runtime issue is where I'm getting at

**Comment by longlho** [comment]
@jasonmit another question: empty string is considered an error (thus falling back) in react-intl. Thoughts? This is basically due to the fact that some localization vendors set untranslated string as empty and/or English so fallback behavior might be a lil non-deterministic.

**Comment by longlho** [comment]
@jasonmit how do I publish both code & docs?

**Reply by jasonmit** [reply]
A standard `npm publish` and the doc publishing should be automatic based on git tagging: https://github.com/ember-intl/ember-intl/actions?query=workflow%3A%22Deploy+Tagged+Docs%22
